### PR TITLE
[System Improvement] Combat log Improvements

### PR DESCRIPTION
**Pull request by Valorith**
## Summary

This PR makes combat log delivery consistent for in-zone group and raid members involved in the same combat event, even when they are outside the normal proximity radius. It also restores attribution for caster-lost cases and fixes combat message paths that bypassed existing filters.

## Rules

New rules (every behavior-changing rule ships **default off** — stock rule_values behave identically to master; operators opt in per rule):

```cpp
RULE_BOOL(Spells, BuffTicAttributionFallback, false, "Keep DoTs ticking and attribute buff tic messages using the buff's stored caster name when the caster mob is dead or gone from zone")
RULE_BOOL(Combat, GroupRaidCombatLogParity, false, "In-zone group/raid members of any party to a combat event receive its combat packets/messages regardless of proximity")
RULE_BOOL(Combat, HealBystanderMessages, false, "Send third-person heal/HoT messages to bystanders and group/raid members")
RULE_BOOL(Combat, DamageShieldSpellAttribution, false, "Send the real spell ID in damage shield OP_Damage packets to bystanders; client versions render nonzero spellid packets differently")
RULE_INT(Range, GroupRaidCombatMessages, 0, "Max distance for group/raid combat log parity delivery; 0 = entire zone (members are enumerated directly, no zone scan)")
RULE_INT(Range, DamageShieldMessages, 200, "The packet range in which damage shield damage packets are sent (values above Range:MobCloseScanDistance are clamped to it)")
RULE_INT(Range, CriticalSpellMessages, 100, "The packet range in which spell crit blast/heal messages are sent (values above Range:MobCloseScanDistance are clamped to it)")
```

Relevant existing rules:

```cpp
RULE_INT(Range, DamageMessages, 50, "The packet range in which damage messages are sent (non-crit)")
RULE_INT(Range, SpellMessages, 75, "The packet range in which spell damage messages are sent")
RULE_INT(Range, CriticalDamage, 80, "The packet range in which critical hit messages are sent")
RULE_INT(Range, MobCloseScanDistance, 600, "Close scan distance")
```

## Implementation

- Adds `EntityList::ForEachCombatLogObserver` plus combat message wrappers for group/raid observer delivery.
- Delivers combat packets and messages to eligible group/raid members beyond the proximity radius without duplicating nearby recipients.
- Applies the observer path to melee, pet, bot, spell damage, spell-landed text, DoT text, damage shield packets, `OP_Action`, and melee crit messages.
- Keeps DoT ticks attributed through stored `caster_name` after the caster dies or zones, while preserving live-caster requirements for hate, focus, and resource taps.
- Adds third-person heal/HoT bystander messages and named fallback behavior for null-caster heals.

## Bug fixes

- Bot finishing blow, headshot, and assassinate now honor `Range:CriticalDamage` and `FilterMeleeCrits`.
- `OTHER_CRIT_HEAL` now honors `FilterSpellCrits`.
- Fixes a null-attacker dereference in the CommonDamage bystander branch.
- Fixes a tautological heal filter check and an unguarded `CastToClient()` in the null-caster heal path.

## Testing

- Built and deployed the zone target locally via the AkkStack Linux/Docker build used for runtime testing.
- Ran grouped and raid in-zone parity tests with `Combat:GroupRaidCombatLogParity=true`, `Range:GroupRaidCombatMessages=0`, and heal threshold `0`.
- Verified melee, misses/defensive results, direct spell damage, DoT ticks, incoming DoTs, and heal bystander messages relay to eligible group/raid members.
- Verified finite range behavior, zone-wide range behavior, close-range duplicate suppression, and that non-group/non-raid bystanders do not receive relayed combat.
## Reviewer notes

- Keep `Combat:DamageShieldSpellAttribution` default-off until M16 verifies damage-shield rendering on supported clients.
- Reverse damage shields still send no bystander packet; that pre-existing TODO is intentionally out of scope.
- Observer filters intentionally mirror existing proximity path semantics, including `FilterShowGroupOnly` behavior, so near and far delivery stay aligned.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Large combat hot-path refactor with many packet/message call sites; behavior only changes when rules are enabled, but misconfiguration or observer dedup bugs could affect zone-wide combat visibility.
> 
> **Overview**
> Adds **opt-in** combat-log behavior (new rules default **off**) so in-zone **group/raid** members tied to a fight can receive the same combat packets and formatted messages **beyond normal proximity**, without double-sending nearby clients. Delivery is routed through new `EntityList` helpers (`ForEachCombatLogObserver`, `QueueCombatClients`, `FilteredMessageCombatString` / `FilteredMessageCombatClose`) and replaces many `QueueCloseClients` / `FilteredMessageCloseString` call sites across melee, spells, DoTs, crits, bots, and damage shields.
> 
> Persists **`caster_is_client`** on `character_buffs` (migration **9352**) and uses stored **`caster_name`** when **`Spells:BuffTicAttributionFallback`** is enabled so **orphaned DoT/HoT ticks** still deal damage, show tic text, and apply **PvP mitigation** after the caster mob is gone; live caster is still required for hate, focus, and resource tap. **`HealDamage`** gains optional fallback naming and **`Combat:HealBystanderMessages`** for third-person heal/HoT lines. Optional **`Combat:DamageShieldSpellAttribution`** and new range rules cap damage-shield and spell-crit broadcast distance.
> 
> Also fixes several combat-message edge cases (null attacker in bystander damage paths, heal filter/`CastToClient` guards, bot null-source buff tics, crit/filter range on bot finishers and spell crit heals) and adds repository/migration unit tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0735685ddc22e098581cc6f40f3ed2c1c9b87b59. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Improved combat message delivery with combat-party-aware routing for group/raid and nearby participants.
  * Added configurable rules controlling buff/DoT attribution fallback, combat-log delivery parity, and delivery ranges for damage shield and critical spell messages.
  * Enhanced observer/combat-log attribution to use stored caster info when the original caster is unavailable.
* **Bug Fixes**
  * Orphaned DoT/HoT ticks now continue processing correctly when fallback attribution is enabled.
  * Critical and bystander combat messages now apply consistent filtering and distance limits.
* **Chores**
  * Added database migration and coverage for new character buffs caster connectivity tracking.

### CODE REVIEW

File: zone/entity.cpp [hunk 1]
```diff
@@ -1793,6 +1793,272 @@ void EntityList::QueueCloseClients(
...
+		}
+
+		if (!candidate->Connected()) {
+			return;
+		}
+
+		float dist_sq = DistanceSquared(candidate->GetPosition(), sender->GetPosition());
+		if (dist_sq <= proximity_sq) {
+			return; // the proximity send already covered (or filtered) them
+		}
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp:2509]: Fixed in daa135d: `ForEachCombatLogObserver` now mirrors the paired proximity sender's inclusive/exclusive range check. `QueueCombatClients` uses strict `<` dedup so clients exactly at the `QueueCloseClients` boundary still receive observer delivery, while the filtered close-message paths keep inclusive dedup to avoid duplicate messages.

---
<sub>Codex</sub>

File: zone/zonedb.cpp [hunk 2]
```diff
@@ -2733,7 +2733,9 @@ void ZoneDatabase::LoadBuffs(Client *client)
 			buffs[e.slot_id].casterid = 0;
 			buffs[e.slot_id].client   = false;
 
-			strncpy(buffs[e.slot_id].caster_name, "", 64);
+			// keep the stored name so buff tic attribution survives the
+			// caster being gone (Spells:BuffTicAttributionFallback)
+			strncpy(buffs[e.slot_id].caster_name, e.caster_name.c_str(), 64);
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/zonedb.cpp]: Fixed in daa135d: these caster-name copies now use `strn0cpy` with `sizeof(...)`, so the fixed buffer is always NUL-terminated.

---
<sub>Codex</sub>

File: zone/spell_effects.cpp [hunk 3]
```diff
@@ -3914,21 +3914,27 @@ void Mob::DoBuffTic(const Buffs_Struct &buff, int slot, Mob *caster)
...
-				effect_value = caster->GetActDoTDamage(buff.spellid, effect_value, this);
+					effect_value = caster->GetActDoTDamage(buff.spellid, effect_value, this);
+
+					caster->ResourceTap(-effect_value, buff.spellid);
+				}
 
-				caster->ResourceTap(-effect_value, buff.spellid);
 				effect_value = -effect_value;
 				Damage(caster, effect_value, buff.spellid, spell.skill, false, i, true);
 			} else if (effect_value > 0) {
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/spell_effects.cpp:3940]: Fixed in daa135d: `Bot::Damage` now allows null-attacker buff-tic damage through to `CommonDamage`, while guarding `EVENT_ATTACK` handling on a real attacker. That keeps orphaned DoT damage applying to bot targets without adding hate/focus/tap work for missing casters.

---
<sub>Codex</sub>

File: zone/attack.cpp [hunk 4]
```diff
@@ -3381,9 +3381,19 @@ void Mob::DamageShield(Mob* attacker, bool spell_ds) {
 		b->target  = attacker->GetID();
 		b->source  = GetID();
 		b->type    = spellbonuses.DamageShieldType;
-		b->spellid = 0x0;
+		// real id only for true buff damage shields: the spell_ds id is a
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/attack.cpp:3396]: Fixed in f1ecaa8: `NPC::Damage` now only fires and resets the attack-event timer when there is a real attacker, so orphaned DoT ticks no longer pass a null attacker into `EventMercNPC`. I also applied the same timer guard to `Bot::Damage` so null-attacker buff ticks do not suppress a later real attacker event. Verified with `cmake --build build-codex --target zone --config RelWithDebInfo -- /m:4`.

---
<sub>Codex</sub>

File: zone/attack.cpp [hunk 5]
```diff
@@ -4750,39 +4767,57 @@ void Mob::CommonDamage(Mob* attacker, int64 &damage, const uint16 spell_id, cons
...
+			const char *caster_name = nullptr;
+			if (attacker) {
+				caster_name = attacker->GetCleanName();
+			} else if (
+				RuleB(Spells, BuffTicAttributionFallback) &&
+				buffslot >= 0 &&
+				buffslot < GetMaxTotalSlots() &&
+				buffs[buffslot].caster_name[0]
+			) {
+				caster_name = buffs[buffslot].caster_name;
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/attack.cpp]: Fixed in a0c3f01: `DoBuffTic` now passes the actual buff array slot to `Damage` for CurrentHP DoT ticks and Bard AEDoT ticks, so `CommonDamage` reads `buffs[buffslot].caster_name` from the exact ticking buff instead of an effect index. Verified with `cmake --build build-codex --target zone --config RelWithDebInfo -- /m:4`.

---
<sub>Codex</sub>

File: zone/entity.h [hunk 6]
```diff
@@ -427,6 +428,39 @@ class EntityList
 	void	RemoveFromAutoXTargets(Mob* mob);
 	void	ReplaceWithTarget(Mob* pOldMob, Mob*pNewTarget);
 	void	QueueCloseClients(Mob* sender, const EQApplicationPacket* app, bool ignore_sender=false, float distance=200, Mob* skipped_mob = 0, bool is_ack_required = true, eqFilterType filter=FilterNone);
+
+	// Combat log group/raid parity (Combat:GroupRaidCombatLogParity): resolves
+	// the parties to a combat event (sender/other, or their owners for pets)
+	// and invokes fn for each in-zone group/raid member beyond proximity_range
+	// of the sender - members within range already received (or filtered) the
+	// proximity copy. Combat-only; non-combat proximity messaging must keep
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.h]: Fixed in 25183b8: the `ForEachCombatLogObserver` comment now describes in-zone observers as group/raid members or the ungrouped client party itself, matching the solo anchor processing. Verified with `git diff --check` and `cmake --build build-codex --target zone --config RelWithDebInfo -- /m:4`.

---
<sub>Codex</sub>

File: zone/mob.h [hunk 7]
```diff
@@ -557,7 +557,7 @@ class Mob : public Entity {
 		bool avoidable = true, int8 buffslot = -1, bool iBuffTic = false, eSpecialAttacks special = eSpecialAttacks::None) = 0;
 	void SetHP(int64 hp);
 	inline void SetOOCRegen(int64 new_ooc_regen) { ooc_regen = new_ooc_regen; }
-	virtual void HealDamage(uint64 ammount, Mob* caster = nullptr, uint16 spell_id = SPELL_UNKNOWN);
+	virtual void HealDamage(uint64 ammount, Mob* caster = nullptr, uint16 spell_id = SPELL_UNKNOWN, const char* caster_name_fallback = nullptr);
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/mob.h]: Fixed in 25183b8: corrected the `HealDamage` declaration parameter name from `ammount` to `amount`; the implementation already used `amount`. Verified with `git diff --check` and `cmake --build build-codex --target zone --config RelWithDebInfo -- /m:4`.

---
<sub>Codex</sub>

File: zone/spell_effects.cpp [hunk 8]
```diff
@@ -3969,7 +3975,7 @@ void Mob::DoBuffTic(const Buffs_Struct &buff, int slot, Mob *caster)
 					} else if (!caster->IsClient())
 						AddToHateList(caster, effect_value);
 				}
-				Damage(caster, effect_value, buff.spellid, spell.skill, false, i, true);
+				Damage(caster, effect_value, buff.spellid, spell.skill, false, slot, true);
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/spell_effects.cpp:3978]: Intentionally unchanged: BardAEDot dealing damage with a null caster is pre-existing master behavior (its Damage call was never gated on a live caster), and the rule-off state is required to match master exactly. Spells:BuffTicAttributionFallback restores the CurrentHP path that master *loses* on caster death; it does not remove bard functionality master has always had. Documented as deviation D3 in the verification report.

File: zone/attack.cpp [hunk 9]
```diff
@@ -4851,11 +4886,37 @@ void Mob::HealDamage(uint64 amount, Mob* caster, uint16 spell_id)
...
+					caster->GetCleanName(), GetCleanName(),
+					(unsigned long long) acthealed, spells[spell_id].name);
+			} else {
+				entity_list.FilteredMessageCombatClose(
+					caster, this, /*skipsender*/ true, RuleI(Range, SpellMessages),
+					Chat::NonMelee, FilterSpellDamage, /*skipped_mob*/ this,
+					"%s healed %s for %llu hit points by %s.",
+					caster->GetCleanName(), GetCleanName(),
+					(unsigned long long) acthealed, spells[spell_id].name);
+			}
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/attack.cpp:4971]: Fixed in 3e80f6316: the third-person heal line now derives its name from the live caster or the buff's stored caster name, and anchors observer delivery on the heal target when the caster mob is gone - matching the orphaned DoT path. Zone build verified.

File: zone/entity.cpp [hunk 10]
```diff
@@ -1793,6 +1793,273 @@ void EntityList::QueueCloseClients(
...
+
+		Client *client = mob->CastToClient();
+		if (client == skipped_mob || (skipsender && client == sender)) {
+			continue;
+		}
+
+		if (DistanceSquared(client->GetPosition(), sender->GetPosition()) <= dist2) {
+			client->FilteredMessage(sender, type, filter, buffer);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp:2757]: Addressed in 86b12618d for the group/raid half: the observer skip now also requires the candidate to be in the sender's close-mob cache, so an in-range member missing from a stale cache gets the observer copy instead of nothing (no double possible - the proximity pass only reaches cache members). For non-member bystanders, the cache is the deliberate design: it has the same staleness characteristics as every existing QueueCloseClients combat message, and this path fires per HoT tick so a full client_list scan is exactly what the perf design forbids. Zone build verified.

File: zone/entity.cpp [hunk 11]
```diff
@@ -1793,6 +1793,279 @@ void EntityList::QueueCloseClients(
...
+	);
+
+	ForEachCombatLogObserver(sender, other, dist, true, skipsender, skip, [&](Client *client) {
+		// FilteredMessageString checks the client's filter before allocating
+		client->FilteredMessageString(
+			sender, type, filter, string_id,
+			message1, message2, message3, message4, message5,
+			message6, message7, message8, message9
+		);
+	});
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp]: Fixed in 4ed58c416: ForEachCombatLogObserver now takes proximity_covers_all_in_range - FilteredMessageCombatString's proximity pass (FilteredMessageCloseString) scans every client in the zone, so its observer loop skips ALL in-range candidates again; the cache-membership dedup from 86b12618d now applies only to the wrappers whose proximity pass actually iterates the close-mob cache (QueueCombatClients, FilteredMessageCombatClose). Zone build verified.

**Comment by Valorith** [comment]
This seems stable in local testing. Waiting for the final patch release build to get locked in before i merge this to master.

File: zone/entity.cpp [hunk 12]
```diff
@@ -45,6 +45,361 @@ extern volatile bool is_zone_loaded;
...
+
+	if (source && !CombatLogClientWithinRange(client, source, distance)) {
+		return false;
+	}
+
+	if (target && !CombatLogClientWithinRange(client, target, distance)) {
+		return false;
+	}
+
+	return source || target;
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp]: Fixed in 95dd44935. The dedup predicate now matches the actual `QueueCloseClients` coverage: once a client is reached by the sender-anchored close-mob/range pass, we treat the proximity packet as covering them and avoid adding observer text. The only fallback left is when the OP_Damage source/target IDs cannot be resolved, where the synthesized observer path is still needed.

Validation: `git diff --check` passed and `ninja zone` completed successfully in the Linux build tree.

---
<sub>Codex</sub>

File: zone/entity.cpp [hunk 13]
```diff
@@ -1793,6 +2148,347 @@ void EntityList::QueueCloseClients(
...
 
+// combat log group/raid parity: a combat participant "anchors" observer
+// delivery to its own client/bot/merc identity, or to its owner for pets and
+// swarm pets; plain NPCs have no anchor
+Mob *EntityList::ResolveCombatLogAnchor(Mob *mob)
+{
+	if (mob->IsOfClientBotMerc()) {
+		return mob;
+	}
+
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp:2306]: Fixed in badd800bc. `ResolveCombatLogAnchor` now returns `nullptr` immediately for a null mob, so `CombatLogFilterAllowsClient` can safely evaluate optional combat parties.

Validation: `git diff --check` passed and `ninja zone` completed successfully in the Linux build tree.

---
<sub>Codex</sub>

File: zone/entity.cpp [hunk 14]
```diff
@@ -1793,6 +2148,347 @@ void EntityList::QueueCloseClients(
...
+		Client *client = mob->CastToClient();
+		if (client == skipped_mob || (skipsender && client == sender)) {
+			continue;
+		}
+
+		if (DistanceSquared(client->GetPosition(), sender->GetPosition()) <= dist2) {
+			if (CombatLogBystanderMessageFilterAllowsClient(client, filter)) {
+				client->Message(type, "%s", buffer);
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp]: Fixed in badd800bc. `FilteredMessageCombatClose` now uses the same `FilteredMessageCheck(sender, filter)` semantics before sending composed heal/HoT bystander text, so `FilterShowSelfOnly` and `FilterShowGroupOnly` behave like the existing filtered proximity path instead of only suppressing `FilterHide`.

Validation: `git diff --check` passed and `ninja zone` completed successfully in the Linux build tree.

---
<sub>Codex</sub>

File: zone/entity.cpp [hunk 15]
```diff
@@ -45,6 +45,345 @@ extern volatile bool is_zone_loaded;
...
+		return true;
+	}
+
+	if (
+		source &&
+		source->IsOfClientBot() &&
+		damage_packet->type == DamageTypeSpell &&
+		IsValidSpell(damage_packet->spellid)
+	) {
+		return true;
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp]: Fixed in 5a2e228c5 by splitting observer damage handling into explicit stop vs packet-fallback outcomes. Client/bot spell OP_Damage with a valid spell id now falls through to the original packet parity delivery instead of suppressing it. Validated with git diff --check and 
inja zone.

---
<sub>Codex</sub>

File: zone/entity.cpp [hunk 16]
```diff
@@ -1793,6 +2132,351 @@ void EntityList::QueueCloseClients(
...
+		false,
+		skipsender,
+		skipped_mob,
+		nullptr,
+		[&](Client *client) {
+			return CombatLogBystanderMessageFilterAllowsClient(client, sender, filter);
+		},
+		[&](Client *client) {
+			if (CombatLogBystanderMessageFilterAllowsClient(client, sender, filter)) {
+				client->Message(type, "%s", buffer);
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp]: Fixed in 5a2e228c5. The parity leg of FilteredMessageCombatClose now checks the paired combat filter against both sender and other/victim, while the proximity dedup check remains aligned with the existing sender-based close send. Validated with git diff --check and 
inja zone.

---
<sub>Codex</sub>

File: zone/entity.cpp [hunk 17]
```diff
@@ -1793,6 +2132,351 @@ void EntityList::QueueCloseClients(
...
+		other,
+		dist,
+		true,
+		true,
+		skipsender,
+		skip,
+		nullptr,
+		[&](Client *client) {
+			return client && client->FilteredMessageCheck(sender, filter);
+		},
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp:2686]: Fixed in 5a2e228c5. String parity delivery now checks CombatLogFilterAllowsClient against both combat parties and then sends the formatted string directly, avoiding FilteredMessageString's sender-only filter gate for parity observers. Validated with git diff --check and 
inja zone.

---
<sub>Codex</sub>

File: zone/zonedb.cpp [hunk 18]
```diff
@@ -2728,12 +2728,14 @@ void ZoneDatabase::LoadBuffs(Client *client)
...
-			strncpy(buffs[e.slot_id].caster_name, c->GetName(), 64);
+			strn0cpy(buffs[e.slot_id].caster_name, c->GetName(), sizeof(buffs[e.slot_id].caster_name));
 		} else {
 			buffs[e.slot_id].casterid = 0;
 			buffs[e.slot_id].client   = false;
 
-			strncpy(buffs[e.slot_id].caster_name, "", 64);
+			// keep the stored name so buff tic attribution survives the
+			// caster being gone (Spells:BuffTicAttributionFallback)
+			strn0cpy(buffs[e.slot_id].caster_name, e.caster_name.c_str(), sizeof(buffs[e.slot_id].caster_name));
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/zonedb.cpp]: Fixed in b500a3010. When the persisted caster is absent, LoadBuffs now restores caster_name only when Spells:BuffTicAttributionFallback is enabled; otherwise it explicitly clears the name, preserving the default-off packet behavior. The persisted client provenance remains available for PvP/death semantics without exposing the caster name. Validated with the Windows RelWithDebInfo zone build.

---
<sub>Codex</sub>

File: zone/bot.cpp [hunk 19]
```diff
@@ -4977,23 +4979,23 @@ void Bot::Damage(Mob *from, int64 damage, uint16 spell_id, EQ::skills::SkillType
...
 	// if spell is lifetap add hp to the caster
-	if (IsValidSpell(spell_id) && IsLifetapSpell(spell_id)) {
+	if (from && IsValidSpell(spell_id) && IsLifetapSpell(spell_id)) {
 		int64 healed = GetActSpellHealing(spell_id, damage);
 		LogCombatDetail("Applying lifetap heal of [{}] to [{}]", healed, GetCleanName());
 		HealDamage(healed);
-		if (from) {
-			entity_list.FilteredMessageClose(this, true, RuleI(Range, SpellMessages), Chat::Emote, FilterSocials, "%s beams a smile at %s", GetCleanName(), from->GetCleanName());
-		}
+		entity_list.FilteredMessageClose(this, true, RuleI(Range, SpellMessages), Chat::Emote, FilterSocials, "%s beams a smile at %s", GetCleanName(), from->GetCleanName());
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/bot.cpp]: Fixed in f99672a64. I removed the duplicate lifetap block from Bot::Damage, leaving Mob::CommonDamage as the single canonical path so the attacker is healed once and the message participants are correct. Validated with the Windows RelWithDebInfo zone build.

---
<sub>Codex</sub>

File: zone/entity.cpp [hunk 20]
```diff
@@ -35,16 +35,505 @@
...
+		return context;
+	}
+
+	if (!context.target && !context.source) {
+		return context;
+	}
+
+	context.delivery = CombatDamageObserverDelivery::Message;
+	context.chat_type = GetCombatDamageChatType(context.damage_packet->type, context.damage, filter);
+	return context;
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp:449]: I checked this against the observer-delivery design and did not change it. Distant parity observers may not have NPC, pet, or mercenary source/target entities in their spawn set, so forwarding native OP_Damage IDs can produce no usable log line. The synthesized text path is the deliberate reliable fallback for those sources; the client/bot exception was added in 5a2e228c5 specifically to prevent valid client/bot spell packets from being dropped. Broadening that exception would regress distant visibility.

---
<sub>Codex</sub>

**Comment by Valorith** [comment]
Addressed the independent `HealDamage` note from [this review](https://github.com/Valorith/Server/pull/149#pullrequestreview-4667589999) in f99672a64. Stored caster-name attribution is now accepted only for genuine buff ticks while `Spells:BuffTicAttributionFallback` is enabled. Normal null-caster healing retains its legacy generic target message, while the explicit `is_buff_tic` context prevents direct/lifetap heals from being classified or broadcast as HoTs. The duplicate Bot lifetap path was also removed. Validated with the Windows RelWithDebInfo zone build.

---
<sub>Codex</sub>

File: zone/entity.cpp [hunk 21]
```diff
@@ -1758,39 +2247,529 @@ void EntityList::QueueCloseClients(
...
+				solos[i] = client;
+			}
+		} else if (anchors[i]->IsBot()) {
+			raids[i] = anchors[i]->CastToBot()->GetStoredRaid();
+			if (!raids[i] && anchors[i]->IsGrouped()) {
+				groups[i] = GetGroupByMob(anchors[i]);
+			}
+		} else if (anchors[i]->IsGrouped()) {
+			groups[i] = GetGroupByMob(anchors[i]);
+		}
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/entity.cpp:2400]: Validated. An ungrouped bot previously produced no party or solo recipient. Commit `0735685dd` now falls back to the bot's in-zone client owner only when the bot has no raid/group, and collapses same-owner pairs to prevent duplicate delivery. The Windows RelWithDebInfo `zone` build and all 144 tests pass.

---
<sub>Codex</sub>
